sql: add RevertSpans and RevertSpansFanout

stevendanna opened the pull request:
This PR adds RevertSpans and RevertSpansFanout.

RevertSpans de-duplicates some code between import rollback and streaming cutover.

Along the way, it fixes a small bug that existed in RevertTables: Only a single RevertRangeRequest is permitted in a batch since RevertRangeRequest has the isAlone flag set. As a result, a future caller of RevertTables would have encountered a fatal error from KV.

RevertSpansFanout uses DistSQL's PartitionSpans to manually fan out multiple RevertRange requests. Since all users of revert range request currently set a limit on the number of keys touched, dist sender doesn't fanout such request.

Epic: none

Release note: None

stevendanna committed:
sql: add RevertSpans and RevertSpansFanout

This PR adds RevertSpans and RevertSpansFanout.

RevertSpans de-duplicates some code between import rollback and
streaming cutover.

Along the way, it fixes a small bug that existed in RevertTables: Only
a single RevertRangeRequest is permitted in a batch since
RevertRangeRequest has the isAlone flag set. As a result, a future
caller of RevertTables would have encountered a fatal error from KV.

RevertSpansFanout uses DistSQL's PartitionSpans to manually fan out
multiple RevertRange requests. Since all users of revert range request
currently set a limit on the number of keys touched, dist sender
doesn't fanout such request.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97845)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
> Reviewed 3 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97845)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msbutler, @rhu713, @rytaft, and @stevendanna)

___
*[`pkg/sql/revert.go` line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97845#-NPU1ADh1GiSmwrd5Gdi:-NPU1ADh1GiSmwrd5Gdj:bhf57b4) ([raw file](https://github.com/cockroachdb/cockroach/blob/40db612acf5a4776f546de0443754511f66aa740/pkg/sql/revert.go#L49)):*
> ```Go
> var maxRevertSpanNumWorkers = settings.RegisterIntSetting(
> 	settings.TenantWritable,
> 	"max_revert_span_parallelism",
> ```

nit: we usually have several parts in a cluster setting name, delimited by periods, with the first part being about the layer, so perhaps name it something like `sql.revert.max_span_parallelism`.
___
*[`pkg/sql/revert.go` line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97845#-NPU1t012M5JhglGVb8V:-NPU1t012M5JhglGVb8W:bmsphm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/40db612acf5a4776f546de0443754511f66aa740/pkg/sql/revert.go#L51)):*
> ```Go
> 	"max_revert_span_parallelism",
> 	"the maximum number of workers used to issue RevertRange request",
> 	8)
> ```

Do we want to add `NonNegativeInt` or `PositiveInt` validation?
___
*[`pkg/sql/revert.go` line 140 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97845#-NPU27M01_8AZAr0qjax:-NPU27M01_8AZAr0qjay:bbfqr7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/40db612acf5a4776f546de0443754511f66aa740/pkg/sql/revert.go#L140)):*
> ```Go
> 		workerPartitions = spanPartitions
> 	} else {
> 		workerPartitions = make([]SpanPartition, maxWorkerCount)
> ```

If we allow the setting to be set to 0, then we need to handle zero case differently.
___
*[`pkg/sql/revert.go` line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97845#-NPU2A5GDwO7n20Oq12O:-NPU2A5GDwO7n20Oq12P:b1omfw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/40db612acf5a4776f546de0443754511f66aa740/pkg/sql/revert.go#L150)):*
> ```Go
> 	progressWorker := func(ctx context.Context) error {
> 		for sp := range completedSpanCh {
> 			if onCompletedCallback != nil {
> ```

nit: perhaps take the `if` out of the loop and perhaps just short-circuit if the calback is nil.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97845)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msbutler, @rhu713, and @yuzefovich)

___
*[`pkg/sql/revert.go` line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97845#-NPU1t012M5JhglGVb8V:-NPXAvSCB8DIEe04G4di:b-koh9jn) ([raw file](https://github.com/cockroachdb/cockroach/blob/40db612acf5a4776f546de0443754511f66aa740/pkg/sql/revert.go#L51)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Do we want to add `NonNegativeInt` or `PositiveInt` validation?
</blockquote></details>

Thanks! PositiveInt should do.
___
*[`pkg/sql/revert.go` line 140 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97845#-NPU27M01_8AZAr0qjax:-NPXAyTb9oREybFmHZDU:b-cqmwfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/40db612acf5a4776f546de0443754511f66aa740/pkg/sql/revert.go#L140)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

If we allow the setting to be set to 0, then we need to handle zero case differently.
</blockquote></details>

Added the PositiveInt validation so the zero case should be OK now.
___
*[`pkg/sql/revert.go` line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97845#-NPU2A5GDwO7n20Oq12O:-NPXB2EMDYzlvF3nql9V:bz34xpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/40db612acf5a4776f546de0443754511f66aa740/pkg/sql/revert.go#L150)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps take the `if` out of the loop and perhaps just short-circuit if the calback is nil.
</blockquote></details>

While addressing this comment, I noticed that I wasn't really handling the error return from this callback correctly.  So, perhaps I've gone a bit overboard, but I've done a few things:

1) short-circuit the 1 worker case to avoid all planning and just call RevertSpans
2) short-circuit the case of a nil onCompletedCallback to avoid the extra layer of goroutines altogether
3) Added an errCh that workers get the last onCompletedCallback error back from.


<!-- Sent from Reviewable.io -->

msbutler commented on pkg/sql/revert.go:
```diff
@@ -94,21 +219,38 @@ func RevertTables(
+			span = resp.ResumeSpan
+
+			if resp.ResumeSpan != nil {
+				if !resp.ResumeSpan.Valid() {
```
nit: perhaps we add some logging around resume key spans and the number of attempts to revert the original span?

stevendanna replied:
We have a VEventf above that can be added on each revert range call.  Note that a resume span is absolutely expected.

msbutler commented on pkg/sql/revert_test.go:
```diff
@@ -106,6 +111,206 @@ func TestTableRollback(t *testing.T) {
+	// the gaps made between previous rows, edit a large swath of rows and add a
+	// large swath of new rows as well.
+	db.Exec(t, "DELETE FROM test WHERE k % 5 = 2")
+	db.Exec(t, "INSERT INTO test (k, rev) SELECT generate_series(10, $1, 10), 10", initRows)
```
was this line suppose to use `extraRows`? It seems like `extraRows` was never used to insert stuff.

stevendanna replied:
Thanks, I meant to parameterize the hard coded 500 below.

msbutler commented on pkg/sql/revert_test.go:
```diff
@@ -106,6 +111,206 @@ func TestTableRollback(t *testing.T) {
+	})
+	t.Run("revert-with-1-worker", func(t *testing.T) {
+		db.Exec(t, "DELETE FROM test WHERE k % 5 = 2")
+		db.Exec(t, "SET CLUSTER SETTING sql.revert.max_span_parallelism = 1")
```
you might need to wait for this cluster setting get fully written, perhaps by using `serverutils.SetClusterSetting()`

stevendanna replied:
I think we are OK in this case. The place that the setting matters is on the node running RevertSpansFanout which is the same place I'm running this query.

msbutler commented on pkg/sql/revert_test.go:
```diff
@@ -106,6 +111,206 @@ func TestTableRollback(t *testing.T) {
+		require.NoError(t,
+			sql.RevertSpansFanout(ctx,
+				kvDB, rsCtx, []roachpb.Span{span}, targetTime, false, 10,
+				func(context.Context, roachpb.Span) error {
```
do you think this test would add coverage if we verified in the callback that the requests were actually distributed at the kv client level?

stevendanna replied:
Always happy to add test coverage, but I'm not actually sure what you mean here. What do you have in mind?

msbutler replied:
I'm actually not sure what i had in mind :D. This test looks fine to me.

msbutler commented on pkg/sql/revert.go:
perhaps in a follow up PR: to maximize parallelism over time, we may want to distribute work to the workers over a queue as they work, instead of pre allocating work, the workers should merely pull the next range of work of some queue when they're done. Perhaps this isn't that big of a deal, if each range will have the same amount of work, but I'm not sure that's the case.

stevendanna replied:
Yeah, the reason I didn't use _just_ a queue here is part of the idea is that by dividing up the work with PartitionSpans, each worker has a list of spans that more or less goes to the same set of nodes, so we hopefully aren't sending a lot of RevertRange requests to the same node at once.

But we still still look at this for the future.

msbutler reviewed (commented):
this is awesome! eager to see benchmarks!

msbutler commented on pkg/sql/revert.go:
```diff
@@ -42,6 +44,14 @@ var predicateDeleteRangeNumWorkers = settings.RegisterIntSetting(
+	settings.TenantWritable,
+	"sql.revert.max_span_parallelism",
+	"the maximum number of workers used to issue RevertRange request",
+	8,
```
should this scale by the number of nodes in the cluster?

stevendanna commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -494,56 +494,70 @@ func revertToCutoverTimestamp(
+					fmt.Sprintf("starting to cut over to the given timestamp %s", cutoverTimestamp))
+			} else {
+				if streamIngestionProgress.ReplicationStatus == jobspb.ReplicationCuttingOver {
+					return errors.AssertionFailedf("cutover already started but cutover time %s is not eligible for cutover",
```
Let's visit in a follow up. We shouldn't need to in the current structure since the frontier can only move forward (unless there is a bug) and the frontier being _ahead_ of the cutover time is fine.

stevendanna commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -494,56 +494,70 @@ func revertToCutoverTimestamp(
+		cutoverTimestamp      hlc.Timestamp
+		spanToRevert          roachpb.Span
+	)
+	if err := ingestionJob.NoTxn().Update(ctx,
```
Yeah, NoTxn() means "I don't have a txn for you, use your own." The Update is still transactional.

stevendanna commented on pkg/sql/revert.go:
```diff
@@ -94,21 +219,38 @@ func RevertTables(
 			})
-		}
-		b.Header.MaxSpanRequestKeys = batchSize
+			// TODO(ssd): We should probably be setting an
```
Yeah, I think that is mostly right.  I could imagine in the serverless usecase also wanting admission control for tenant streaming cutover which is why I left it as a TODO for now.

stevendanna committed:
sql: use a mutex rather than channels in RevertSpansFanout

Release note: None

stevendanna commented on pkg/sql/revert.go:
:D. Beacuse Erlang rotted my brain, I almost always reach for a channel over a mutex. But, in this case, a Mutex really does clean things up. Nice suggestion.

Almost makes me want to just push concurrency concerns to the caller, but meh, there is only one caller at the moment so I'll leave it for now.

stevendanna commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -701,6 +673,108 @@ func (s *streamIngestionResumer) OnFailOrCancel(
 	})
 }
 
```
If we add a third version of this, I think I agree. As it is, this code is duplicating some code from the index backfiller which uses a similar strategy.

stevendanna committed:
sql: avoid modifying kv request after sending

Under the race detector, kvclient uses a transport that disallows
mutations of of a request send to the transport. This change ensures
we create a new request for each resume span rather than re-using the
old request.

Epic: none

Release note: None

stevendanna committed:
roachtest: use capability to opt tenant out of rate limiting

Epic: none

Release note: None

stevendanna commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -701,6 +673,108 @@ func (s *streamIngestionResumer) OnFailOrCancel(
+	if err := c.job.NoTxn().FractionProgressed(ctx, jobs.FractionUpdater(fractionRangesFinished)); err != nil {
+		return jobs.SimplifyInvalidStatusError(err)
+	}
+	if c.onJobProgressUpdate != nil {
```
Overall, my instinct is often to _not_ tie together dependencies if I don't have to.  It tends to make things a bit easier to unit test if this thing gets more complicated.

stevendanna commented on pkg/sql/revert.go:
```diff
@@ -80,12 +90,127 @@ func RevertTables(
+//
+// We do this to get parallel execution of RevertRange even in the
+// case of a non-zero batch size. DistSender will not parallelize
+// requests with non-zero MaxSpanRequestKeys set.
```
I didn't do this because I'm not sure that is actually ideally what we want here.

stevendanna commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -7568,12 +7568,6 @@ expires until the statement bundle is collected`,
 					filter = kvpb.MVCCFilter_All
 				}
 
-				req := &kvpb.ExportRequest{
```
Oh did it? Let me go check, perhaps I hadn't git-pulled before working on this.

stevendanna commented:
bors r=msbutler

stevendanna commented:
bors retry
